_posts/2018-08-01-gke-multi-stage.md

```diff
@@ -0,0 +1,424 @@
+```
+kubectl -n kube-system logs deployment/cert-manager-cert-manager
+
+Certificate issued successfully
+```
+
+## OpenFaaS production setup
+
+Generate a random password and create the `basic-auth` secret in the openfaas-prod namespace:
+
+```bash
+password=$(head -c 12 /dev/urandom | shasum | cut -d' ' -f1)
+
+kubectl -n openfaas-prod create secret generic basic-auth \
+--from-literal=basic-auth-user=admin \
+--from-literal=basic-auth-password=$password
+```
+
+Create the production configuration (replace `example.com` with your own domain):
+
+{% gist fc8d7ef8f4af3d0d81a9f28ff8c6edcb openfaas-prod.yaml %}
+
+For the production deployment the OpenFaaS gateway has high-availability through two replicas. 
+We make sure those replicas are scheduled on different nodes through the use of a pod anti-affinity rule.
+Note that `operator.createCRD` is set to false since the `functions.openfaas.com` custom resource definition is already present on the cluster.
+
+Save the above file as `openfaas-prod.yaml` and install the OpenFaaS production instance with helm:
+
+```bash
+helm upgrade openfaas-prod --install openfaas/openfaas \
+    --namespace openfaas-prod  \
+    -f openfaas-prod.yaml
+```
+
+## Manage your functions with kubectl
+
+![openfaas-operator](/images/gke-multi-stage/operator.png)
+
+Using the OpenFaaS CRD you can define functions as Kubernetes custom resources.
+
+Create `certinfo.yaml` with the following content:
+
+{% gist fc8d7ef8f4af3d0d81a9f28ff8c6edcb certinfo.yaml %}
+
+In oder to make use of the preemptible node pool you need to 
+set the affinity constraint to `cloud.google.com/gke-preemptible=true` so that certinfo will be deployed on the `fn-pool`.
+
+Use `kubectl` to deploy the function to both environments by changing the namespace parameter:
+
+```bash
+kubectl -n openfaas-stg-fn apply -f certinfo.yaml
+kubectl -n openfaas-prod-fn apply -f certinfo.yaml
+```
+
+The certinfo function will tell you the SSL information for your domain. 
+User certinfo to verify both endpoints are live:
+
+```
+curl -d "openfaas-stg.example.com" https://openfaas-stg.example.com/function/certinfo
+Issuer Let's Encrypt Authority X3
+....
+
+curl -d "openfaas.example.com" https://openfaas.example.com/function/certinfo
+Issuer Let's Encrypt Authority X3
+....
+```
+
+You can get the list of functions running in a namespace:
+
+```
+kubectl -n openfaas-stg-fn get functions
+```
+
+And you can delete a function from an namespace with:
+
+```
+kubectl -n openfaas-stg-fn delete function certinfo
+```
+
+## Developer workflow
+
+When it comes to the development workflow you will be using OpenfaaS CLI.
+
+Install faas-cli and login to the staging instance with:
+
+```bash
+curl -sL https://cli.openfaas.com | sudo sh
+
+echo $stg_password | faas-cli login -u admin --password-stdin \
+ -g https://openfaas-stg.example.com
+```
+
+Using faas-cli, Git and kubectl a development workflow would look like this:
+
+_1._ Create a function using the golang template 
```

Comment:
Perhaps we should put `go` in this line to prevent someone typing `--lang golang`?